Enable Briefcase to package apps that weren't build with Briefcase

freakboy3742 opened the pull request:
Briefcase is not the only tool for converting Python applications into running binaries. However, most of those tools don't provide support for signing and notarization, or packaging as DMG/PKG (on macOS), or MSI (on Windows), or DEB/RPM (on Linux). 

This PR modifies Briefcase to add the concept of an "external app" - an app that wasn't built with Briefcase, but *is* being packaged by Briefcase. This is done by adding a `package_path` configuration item. 

Windows requires a template change to reflect possible deep binary paths:
* beeware/briefcase-windows-app-template#66
* beeware/briefcase-windows-VisualStudio-template#59

Fixes #2326.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented on docs/reference/configuration.rst:
```diff
@@ -331,6 +335,28 @@ compatibility; on the platforms where it *is* used, the interpretation of the va
+
+If ``package_path`` is defined, ``sources`` must *not* be defined.
+
+``package_binary_path``
+~~~~~~~~~~~~~~~~~~~~~~~
```
As discussed in external-apps.rst, this setting might not be necessary. But if it's kept, it should mention that it's only implemented on Windows.

freakboy3742 commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+<macos-entitlement>` in your ``pyproject.toml``.
+
+If Briefcase's configuration options are not sufficient for your requirements, you can
+either fork the template used by Briefcase and use the ``template`` option to point at
```
This is a situation where we hit the limits of Restructured text - because we can't put markup in a link. I'll have to investigate what can be done to make it an `:attr:` (or some other Sphinx reference that renders as a keyword).

mhsmith replied:
If we have to choose between a link and monospace, I think the link is more important.

freakboy3742 replied:
Ok - setting this up was a bigger job than fits into this PR; I've added #2341 as a PR to add references. Which ever one lands first out of that one and this one can then update the couple of extra settings (and setting references) added by this PR.

freakboy3742 commented on docs/reference/configuration.rst:
```diff
@@ -331,6 +335,28 @@ compatibility; on the platforms where it *is* used, the interpretation of the va
+
+This is used to configure Briefcase to :doc:`package an external application
+</how-to/external-apps>`. It is not required if you are using Briefcase for the entire
+app creation process.
```
Agreed about ordering; however, I think we can use ReST markup to draw better attention to it.

freakboy3742 reviewed (commented):
I've applied most of those changes; I still need to investigate 2 things:

* Can we add ReST metadata to allow linking to settings?
* Is `formal_name` actually used on macOS?

Three questions that need a decision:
* What should we do with error handling on `sources`/`requires`/`test-requires`/`cleanup-paths`/...?
* Should we change the name of `package_path`? Candidates are `external_app_path`, `external_package_path`... maybe others?
* Confirm that my answer about `package_binary_path` makes sense

mhsmith commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+---------------------------
+
+The process of configuring an external app is mostly the same as a normal Briefcase app,
+with one important difference: you do *not* define a ``sources`` entry, but instead
```
OK, I guess porting a full Briefcase configuration to an external configuration isn't very likely anyway, so let's leave it the way it is.

mhsmith commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+
+The process of configuring an external app is mostly the same as a normal Briefcase app,
+with one important difference: you do *not* define a ``sources`` entry, but instead
+define a ``package_path``. The ``package_path`` should be the content of your build tool
```
OK, makes sense.

mhsmith commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+to be installed into the end-user's Program Files directory. By default, it is assumed
+that the binary used to launch the app is named ``<package_path>/<formal name>.exe`` (or
+``<package_path>/<formal name>.exe`` for a command-line app). If this is not the name or
+path to your executable, you can define ``package_binary_path`` to define the path
+inside ``package_path`` to your binary. For example, a ``package_path`` value of
+``external``, and a ``package_binary_path`` value of ``somewhere/myapp.exe``, would
```
OK, I see. In that case my only suggestion is to rename it to `package_executable_path`.

mhsmith commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+</reference/commands/package>` for details.
+
+Briefcase will re-generate the templated app content on every invocation of ``briefcase
+package``. On a second invocation of ``briefcase package`` for the same project, you
+will be prompted to overwrite the application. This will only overwrite the templated
+packaging metadata, not the application itself.
```
Looks good, but there's a test failure because of backslashes on Windows.

mhsmith commented on docs/reference/configuration.rst:
```diff
@@ -338,24 +338,32 @@ value is interpreted.
 
+.. admonition:: Only for external apps
+
+    This setting is only required if you're using Briefcase to :doc:`package an external
+    application </how-to/external-apps>`. It is not required if you are using Briefcase
```
This setting is never required.

```suggestion
    This setting is only allowed if you're using Briefcase to :doc:`package an external
    application </how-to/external-apps>`. It is not allowed if you are using Briefcase
```

mhsmith commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+
+On Windows, the ``package_path`` should be the directory that contains all the content
+to be installed into the end-user's Program Files directory. By default, it is assumed
+that the executable used to launch the app is named ``<package_path>/<formal name>.exe`` (or
+``<package_path>/<formal name>.exe`` for a command-line app). If this is not the name or
```
```suggestion
that the executable used to launch the app is named ``<package_path>/<formal_name>.exe`` (or
``<package_path>/<app_name>.exe`` for a :doc:`console app </how-to/cli-apps>`). If this is not the name or
```

freakboy3742 commented on docs/how-to/external-apps.rst:
```diff
@@ -0,0 +1,160 @@
+~~~~~
+
+On macOS, the ``package_path`` should be the location of the ``.app`` bundle that you
+want to distribute to users. It is assumed that the configured ``formal_name`` is the
+name of the app bundle (e.g., ``Hello World.app`` should use a formal name of ``Hello
```
It looks like the formal name *is* used; it's rolled out into the template as part of the PKG installation instructions, and it's used to construct the name of the distribution archive. Those uses could *possibly* be configured to be independent... but given this is a niche use case to start with, and the complexities of explaining the subtle difference between the name of the `package_path` artefact and the `formal_name`... it's probably better to just validate and reject inconsistent `package_path` values. 

Plus, if there really is a use case for having a different formal name on one platform - it's possible to set `formal_name` in a platform configuration.

freakboy3742 commented:
Updates in this pass:

* I've fixed the test failure on Windows (/me shakes fist at path separators)
* I've added validation around `formal_name` on macOS (and updated the documentation to match)
* I've renamed `package_path` to `external_package_path`,
* I've renamed `package_binary_path` to `external_package_executable_path`

The only outstanding item at this point is the cross-referencing of some of those settings, but that turned in to a bigger job, so I've put up #2341 to add the cross-referencing tooling. Once that's landed, we can update this PR (or vice versa).
